caffe2/core/typeid.h

```diff
@@ -352,25 +379,38 @@ class TypeMeta {
 // and as a result, we define these two macros slightly differently.
 
 #ifdef _MSC_VER
-#define CAFFE_KNOWN_TYPE(T)                              \
-  template <>                                            \
-  CAFFE2_EXPORT CaffeTypeId TypeMeta::Id<T>() {          \
-    static bool type_id_bit[1];                          \
-    static TypeNameRegisterer<T> registerer(             \
-        reinterpret_cast<CaffeTypeId>(type_id_bit), #T); \
-    return reinterpret_cast<CaffeTypeId>(type_id_bit);   \
+#define CAFFE_KNOWN_TYPE(T)                                                   \
+  template <>                                                                 \
+  CAFFE2_EXPORT CaffeTypeId TypeMeta::Id<T>() {                                    \
+    static bool type_id_bit[1];                                               \
+    static const CaffeTypeId type_id(reinterpret_cast<intptr_t>(type_id_bit));     \
+    static TypeNameRegisterer<T> registerer(                                  \
+        type_id, #T);                                                         \
+    return type_id;                                                           \
   }
 #else // _MSC_VER
-#define CAFFE_KNOWN_TYPE(T)                              \
-  template <>                                            \
-  CaffeTypeId TypeMeta::Id<T>() {                        \
-    static bool type_id_bit[1];                          \
-    static TypeNameRegisterer<T> registerer(             \
-        reinterpret_cast<CaffeTypeId>(type_id_bit), #T); \
-    return reinterpret_cast<CaffeTypeId>(type_id_bit);   \
+#define CAFFE_KNOWN_TYPE(T)                                                   \
+  template <>                                                                 \
+  CaffeTypeId TypeMeta::Id<T>() {                                                  \
+      static bool type_id_bit[1];                                             \
+    static const CaffeTypeId type_id(reinterpret_cast<intptr_t>(type_id_bit));     \
+    static TypeNameRegisterer<T> registerer(                                  \
+        type_id, #T);                                                         \
+    return type_id;                                                           \
   }
 #endif
 
-} // namespace caffe2
+}
+
+// Define adapters for c10 code
+namespace c10 {
+using TypeId = caffe2::CaffeTypeId;
```

Comment:
this is totally confusing and should not be landed in my opinion. It has different names in different namespace and would be hard to everybody reading the code.

Again: either use caffe2::CaffeTypeId and clean up later by renaming usages or just move TypeId proper to c2 namespace

Reply:
(minor correction about Dima's comment: move TypeId proper to c10 namespace)